ui/apps/platform/src/hooks/usePaginatedQuery.ts

```diff
@@ -0,0 +1,187 @@
+    const [isEndOfResults, setIsEndOfResults] = useState(false);
+    const [lastFetchError, setLastFetchError] = useState<Error | null>(null);
+
+    const [itemKeys, setItemKeys] = useState<Set<ItemKey>>(new Set());
+    const [itemPages, setItemPages] = useState<Item[][]>([]);
+
+    // Pages are zero-indexed, so the length of the current cached data array is equal
+    // to the current page
+    const page = itemPages.length;
+
+    const fetchPageHandler = useCallback(
+        (fetchFn: typeof queryFn, nextPage: number, nextItemKeys: Set<ItemKey>) => {
+            return fetchFn(nextPage)
+                .then((res) => {
+                    setLastFetchError(null);
+                    if (res.length < pageSize) {
+                        setIsEndOfResults(true);
+                    }
+                    const { nextPageItems, nextKeys } = filterNextPage(
+                        res,
+                        nextItemKeys,
+                        dedupKeyFn
+                    );
+                    setItemPages((prevData) => {
+                        const nextData = [...prevData];
+                        // We set directly via array index here instead of appending to prevent bugs when
+                        // additional fetch requests are initiated before the first completes.
+                        nextData[nextPage] = nextPageItems;
+                        return nextData;
+                    });
+                    setItemKeys(nextKeys);
+                })
+                .catch((err) => {
+                    setLastFetchError(err);
+                    if (onError) {
+                        onError(err);
+                    }
+                })
+                .finally(() => {
+                    setIsFetchingNextPage(false);
+                    setIsRefreshingResults(false);
+                });
+        },
+        [dedupKeyFn, onError, pageSize]
+    );
+
+    // Retain a reference to the debounced fetch function and the raw fetch function for cases
+    // we know that we want to fetch the data _now_.
+    // (e.g. We would want to debounce for text input `keypress` events, but a single
+    // "View more" button that is immediately disabled once clicked can fetch without delay.)
+    const pageFetcher = useMemo(() => {
+        setIsFetchingNextPage(true);
+        return {
+            debounced: debounce(fetchPageHandler, debounceRate ?? 0),
+            immediate: fetchPageHandler,
+        };
+    }, [debounceRate, fetchPageHandler]);
+
+    const fetchNextPage = useCallback(
+        (immediate = false) => {
+            if (immediate) {
+                return pageFetcher.immediate(queryFn, page, itemKeys);
+            }
+            return pageFetcher.debounced(queryFn, page, itemKeys);
+        },
+        [pageFetcher, queryFn, page, itemKeys]
+    );
+
+    const clearPages = useCallback(() => {
+        setItemPages([]);
+        setItemKeys(new Set());
```

Comment:
For my understanding, does any state update depend on new set which changes referential identity?

Reply:
Yes, only the `fetchNextPage` function that is wrapped in a `useCallback`. I captures the current value of `itemKeys` in a closure and needs to be replaced each time that value reference changes.